README.md
# YTPrinter

## Buy immediately upon Coin Bureau's video launch

## TODO:
- Decouple trader.py to trader.py and portfolio_manager.py.
    - trader: can initialize margin buys/sells, transfers, future trading. 
    - portfolio_manager: calls trader at the right times. Is in charge of logistics and interfacing with strats.json. 
- strats.json: keeps track of current margin portfolio as well. 
- rename youtube_checker to strategy_manager. 
- in strategy_manager, every time api list reaches end, instead of sleeping for x seconds, call portfolio_manager to do update on saving most recent % of portfolio + updating client 

## SMALL TODOs:
- If in portfolio_manager, put try catch on trader, if trader fails, refresh client, fetch state in portfolio_manager (updates strats to be same as information on binance), then calls trader again
- In strategy strats set leverage to use
- In strategy after a strat is submitted, print a summary (max loss, tp, sl, time stops,etc)
- Discord capabilities. messenger.py

# Past notes:
## Runtime cycle
### 1. Initalize
1. Checking if waiting for CB to post (cross margin account != 0), or what asset is long. Set value to long
- long: bool -- whether currently long (go to (3)) or no position (waiting for CB to post)
- MAIN: What account to park USDT in (ie 1INCHUSDT)
- STRATS: 
    - A dictionary/csv/json pegging keywords to margin account
    - For asset, np.nan, or price bought at, and if active


### 2. Waiting
1. constantly looping over whether CB posted a new video with keywords
2. If so: 
    1. Transfer to pre-set margin account
    2. Buy with max leverage
    3. go to (3), passing in asset to keep track of and price at which it was bought at (save to STRATS)

### 3. Looking to sell
Reads from STRATS to see which strat is active (should only be one, else firt)
1. Set a stop limit SL order if necessary (>$10 of unsecured asset) at 0.5% below buy price (if exists given order), at the most recent stop limit price for this asset if within 3 days (if exists), or as a last resort 0.5% below current price. 
2. Start timer for 1 hour and 3 hours
3. If amount of asset above $10 (still long the position)
    1. If did not sell 5% already: sell if price is above 5% market order for 33% of position
    2. If 1 hour is up, sell 50% or remaming amount (irregardless)
    3. After 3 hours, liquidate everything, go to (4) 
4. ELSE (all liquidated via stop loss): go to (4)
5. If amount of asset <$10 total, then go to (4) 

### 4. Return to normal state
1. Mark STRATS as completed for asset
2. Send USDT to MAIN
3. If STRATS exists strats not touched yet, go to (2), else finish program. 

## Strats
- `finished` if all strats are expended
- `long` either the pair longing or None
- `main` is binance margin account to store USDT when not in use
- `pair[strats['long']]` 
    - ['pos'] is the position of the `long` pair. 0 is default--have not bought, waiting. 1 is have bought and looking to sell. -1 is finished.
    - ['price'] is the buy price   
    - ['tp'] 
        - ['pos'] is 0 if did not take take-profit yet. 1 if already did
        - ['pct'] is take profit level price*(1+pct)
        - ['amt'] is amount of position to take profit. 0.3 == 30%
    - ['sl'] 
        - ['pos'] is 0 if did not take stop-loss yet. 1 if already did
        - ['pct'] is stop-loss level price*(1-pct)
    - ['early'] 
        - ['pos'] is 0 if did not sell early yet, 1 if did sell early
        - ['time'] is time of early time limit
        - ['amt'] is amount to sell 
    - ['late'] 
        - ['pos'] is 0 if did not sell late yet, 1 if did sell late
        - ['time'] is time of late time limit
        - ['amt'] is amount to sell 
    

## Warnings:
- check if it takes time to init a binance client given imports done

## Uhhh
- Centrifuge CFG
- TrueFi TRU
- Goldfinch
- 


past-code/history.txt
{"order": [{"tranId": 72599626942, "clientTag": ""}, {"symbol": "FETUSDT", "orderId": 373623797, "clientOrderId": "PwHrNIuse1SmboCd9LTGYc", "transactTime": 1628886726867, "price": "0", "origQty": "22.2", "executedQty": "22.2", "cummulativeQuoteQty": "11.529546", "status": "FILLED", "timeInForce": "GTC", "type": "MARKET", "side": "SELL", "fills": [{"price": "0.51935", "qty": "21.8", "commission": "0.01132183", "commissionAsset": "USDT"}, {"price": "0.51929", "qty": "0.4", "commission": "0.00020772", "commissionAsset": "USDT"}], "isIsolated": true}, {"symbol": "CELRUSDT", "orderId": 179187721, "clientOrderId": "Xqdu5OW18Ma9asndB8YBxR", "transactTime": 1628886727225, "price": "0", "origQty": "252.4", "executedQty": "252.4", "cummulativeQuoteQty": "11.612924", "status": "FILLED", "timeInForce": "GTC", "type": "MARKET", "side": "BUY", "fills": [{"price": "0.04601", "qty": "252.4", "commission": "0.2524", "commissionAsset": "CELR"}], "isIsolated": true}, {"symbol": "CELRUSDT", "orderId": 179189466, "clientOrderId": "tw7s68hrslm6J14mtodqhT", "transactTime": 1628886965875, "price": "0", "origQty": "252.2", "executedQty": "252.2", "cummulativeQuoteQty": "11.596156", "status": "FILLED", "timeInForce": "GTC", "type": "MARKET", "side": "SELL", "fills": [{"price": "0.04598", "qty": "252.2", "commission": "0.01159616", "commissionAsset": "USDT"}], "isIsolated": true}, {"symbol": "FETUSDT", "orderId": 373627162, "clientOrderId": "aKVmsZZnyUosp136vjnVdW", "transactTime": 1628886966303, "price": "0", "origQty": "22.1", "executedQty": "22.1", "cummulativeQuoteQty": "11.368903", "status": "FILLED", "timeInForce": "GTC", "type": "MARKET", "side": "BUY", "fills": [{"price": "0.51443", "qty": "22.1", "commission": "0.0221", "commissionAsset": "FET"}], "isIsolated": true}, {"tranId": 72601997151, "clientTag": ""}, {"symbol": "CELRUSDT", "orderId": 179195302, "clientOrderId": "IiFfcAHfNiv1lr5R3UyRqI", "transactTime": 1628887989089, "price": "0", "origQty": "260.5", "executedQty": "260.5", "cummulativeQuoteQty": "11.98821", "status": "FILLED", "timeInForce": "GTC", "type": "MARKET", "side": "SELL", "fills": [{"price": "0.04602", "qty": "260.5", "commission": "0.01198821", "commissionAsset": "USDT"}], "isIsolated": true}, {"symbol": "FETUSDT", "orderId": 373636804, "clientOrderId": "62N3LxgesLsL66u8akwAcu", "transactTime": 1628887989683, "price": "0", "origQty": "23.5", "executedQty": "23.5", "cummulativeQuoteQty": "12.006855", "status": "FILLED", "timeInForce": "GTC", "type": "MARKET", "side": "BUY", "fills": [{"price": "0.51093", "qty": "23.5", "commission": "0.0235", "commissionAsset": "FET"}], "isIsolated": true}, {"symbol": "FETUSDT", "orderId": 373638379, "clientOrderId": "LiSMJnsrQRImFNOzyDpSO3", "transactTime": 1628888167062, "price": "0", "origQty": "23.5", "executedQty": "23.5", "cummulativeQuoteQty": "11.967375", "status": "FILLED", "timeInForce": "GTC", "type": "MARKET", "side": "SELL", "fills": [{"price": "0.50925", "qty": "23.5", "commission": "0.01196738", "commissionAsset": "USDT"}], "isIsolated": true}, {"symbol": "CELRUSDT", "orderId": 179196191, "clientOrderId": "WAG7r8ZUWDVkmOf0gRYSD7", "transactTime": 1628888167485, "price": "0", "origQty": "260.7", "executedQty": "260.7", "cummulativeQuoteQty": "12.00784", "status": "FILLED", "timeInForce": "GTC", "type": "MARKET", "side": "BUY", "fills": [{"price": "0.04605", "qty": "0.2", "commission": "0.0002", "commissionAsset": "CELR"}, {"price": "0.04606", "qty": "260.5", "commission": "0.2605", "commissionAsset": "CELR"}], "isIsolated": true}, {"tranId": 72602300126, "clientTag": ""}, {"tranId": 72603150921, "clientTag": ""}, {"symbol": "FETUSDT", "orderId": 373643514, "clientOrderId": "kxF2fp0CqOMqsjeKfOwEQq", "transactTime": 1628888650193, "price": "0", "origQty": "23.4", "executedQty": "23.4", "cummulativeQuoteQty": "11.919492", "status": "FILLED", "timeInForce": "GTC", "type": "MARKET", "side": "SELL", "fills": [{"price": "0.50938", "qty": "23.4", "commission": "0.01191949", "commissionAsset": "USDT"}], "isIsolated": true}, {"symbol": "CELRUSDT", "orderId": 179198926, "clientOrderId": "vObpBuBo1ihbhLql4I7uaG", "transactTime": 1628888650746, "price": "0", "origQty": "259.7", "executedQty": "259.7", "cummulativeQuoteQty": "11.948797", "status": "FILLED", "timeInForce": "GTC", "type": "MARKET", "side": "BUY", "fills": [{"price": "0.04601", "qty": "259.7", "commission": "0.2597", "commissionAsset": "CELR"}], "isIsolated": true}, {"symbol": "CELRUSDT", "orderId": 179199495, "clientOrderId": "8o4uDatXlprcZPwOhQXOiX", "transactTime": 1628888767719, "price": "0", "origQty": "259.5", "executedQty": "259.5", "cummulativeQuoteQty": "11.92662", "status": "FILLED", "timeInForce": "GTC", "type": "MARKET", "side": "SELL", "fills": [{"price": "0.04596", "qty": "259.5", "commission": "0.01192662", "commissionAsset": "USDT"}], "isIsolated": true}, {"symbol": "FETUSDT", "orderId": 373643835, "clientOrderId": "qVGZ37380uKHoUrUJa9jNl", "transactTime": 1628888768125, "price": "0", "origQty": "23.5", "executedQty": "23.5", "cummulativeQuoteQty": "11.985", "status": "FILLED", "timeInForce": "GTC", "type": "MARKET", "side": "BUY", "fills": [{"price": "0.51", "qty": "23.5", "commission": "0.0235", "commissionAsset": "FET"}], "isIsolated": true}, {"tranId": 72603308100, "clientTag": ""}, {"tranId": 72603772730, "clientTag": ""}, {"symbol": "CELRUSDT", "orderId": 179200639, "clientOrderId": "Y594MJECPC5qWliTh5YfLO", "transactTime": 1628889127962, "price": "0", "origQty": "259.3", "executedQty": "259.3", "cummulativeQuoteQty": "11.935579", "status": "FILLED", "timeInForce": "GTC", "type": "MARKET", "side": "SELL", "fills": [{"price": "0.04603", "qty": "259.3", "commission": "0.01193558", "commissionAsset": "USDT"}], "isIsolated": true}, {"symbol": "FETUSDT", "orderId": 373645153, "clientOrderId": "nDBEInZ8gErfIxDK5vzTjF", "transactTime": 1628889128165, "price": "0", "origQty": "23.3", "executedQty": "23.3", "cummulativeQuoteQty": "11.929367", "status": "FILLED", "timeInForce"

strategy.py
import json
import os

STRATS = 'strats.json'
# Strats must have:

def read_strat():
    """
    reads and returns strat from json. Also processes strats filling in nan values with appropriate values 
    """
    try:
        with open(STRATS, 'r') as f:
            data = json.load(f)
        return data
    except:
        print("Warning: Error in opening strat, returning empty json")
        return {}

def save_strat(strat):
    """
    takes in json of the strat and saves it
    """
    os.remove(STRATS)
    with open(STRATS, 'w') as f:
        json.dump(strat, f, indent=4)

def proofread_strat(strats):
    """
    proofreads strat and makes sure everything is in agreement, no contradictions in the json file
    ie, if all strats have -1, then change file to finished
    """
    if strats['pair']: #Not empty
        strats['finished'] = 0 if (False in list(map(lambda x: strats['pair'][x]['pos']==-1, strats['pair']))) else 1
    else: strats['finished'] = 1
    if strats['finished']: strat['long'] = None
    return strats

def json_editor():
    """
    if ran as a script, provides terminal interface to edit strats
    """
    active = True
    while active:
        strats = read_strat()
        cmd = input("Type [v] to view all strats, [a] to add strat, [d] to delete strat, [i] to init strats, [q] to quit\n")
        if (cmd.lower() == "v"): view_all_strats(strats)
        elif (cmd.lower() == "i"): init_strats(strats)
        elif (cmd.lower() == "a"): add_strat(strats)
        elif (cmd.lower() == "d"): delete_strat(strats)
        elif (cmd.lower() == "q"): active = False
        else: print("Invalid command...")

def init_strats(strats):
    """
    properly inits strat to have bare minimum if given empty file, else ignores file
    """
    try: strats['finished']
    except: strats['finished'] = 0
    try: strats['long']
    except: strats['long'] = None
    try: strats['pair']
    except: strats['pair'] = {}
    save_strat(strats)

def view_all_strats(strats):
    """
    views all strats. Throws errors if not properly initalized
    """
    print("="*60)
    print(f"Finished = {strats['finished']}")
    print(f"Long     = {strats['long']}")
    for strat in strats['pair']:
        info = strats['pair'][strat]
        print("~"*60)
        print(f"{strat}: pos = {info['pos']}")
        print("Keywords:")
        for keyword in info['keywords']:
            print(f"- {keyword}")

        print(f"tp:     pos = {info['tp']['pos']}         pct = {info['tp']['pct']}           amt = {info['tp']['amt']}")
        print(f"sl:     pos = {info['sl']['pos']}         pct = {info['sl']['pct']}")
        print(f"early:  pos = {info['early']['pos']}         time = {int(info['early']['time']/60)} min         amt = {info['early']['amt']}")
        print(f"late:   pos = {info['late']['pos']}         time = {int(info['late']['time']/60)} min")

def add_strat(strats):
    """
    prompt user to add a strat and saves
    """
    asset = str(input("Enter name of asset to long (ie BTCUSDT)\n")).upper()
    keywords = []
    while True:
        word = str(input("Enter a keyword. [q] to finish. [r] to restart\n")).lower()
        if word=="q": break
        if word=="r": keywords = []
        else:
            keywords.append(word)
            print(f"Keywords: {keywords}")
    tp_pct = float(input("Enter take profit price multiple (default 1.05)\n"))
    tp_amt = float(input("Enter take profit amount (default 0.3)\n"))
    sl_pct = float(input("Enter stop loss price multiple (default 0.995)\n"))
    early_time = int(input("Enter whole minutes for early time-stop (default 60)\n"))*60
    early_amt = float(input("Enter early time-stop amount (default 0.5)\n"))
    late_time = int(input("Enter whole minutes for late time-stop (default 180)\n"))*60
    tp = {'pos':0, 'pct':tp_pct, 'amt':tp_amt}
    sl = {'pos':0, 'pct':sl_pct}
    early = {'pos':0, 'time':early_time, 'amt':early_amt}
    late = {'pos':0, 'time':late_time, 'amt':1.0}
    strats['pair'][asset] = {'pos':0, 'keywords':keywords, 'tp':tp, 'sl':sl, 'early':early, 'late':late}
    strats['finished'] = 0
    save_strat(strats)
    print(f"{asset} strat successfully added")



def delete_strat(strats):
    """
    prompts user to delete a strat and saves
    """
    strat = str(input("Enter name of strat to delete. [q] to quit. [v] to view all strats first\n")).lower()
    if strat=="q": return
    if strat=="v": view_all_strats(strats)
    strat = strat.upper()
    if strat in strats['pair']:
        if str(input(f"Are you sure you want to delete {strat}? [Y] to continue\n"))=="Y":
            del strats['pair'][strat]
            print(f"Successfully deleted strat {strat}")
        else: 
            print(f"Continuing...")
    else:
        print(f"Strat {strat} does not exist")

    strats = proofread_strat(strats)
    save_strat(strats)

if __name__ == "__main__":
    json_editor()












